cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -970,6 +970,78 @@ class AzureFileSystem::Impl {
     return stream;
   }
 
+ private:
+  Status DeleteDirContentsWihtoutHierarchicalNamespace(const AzureLocation& location,
+                                                       bool missing_dir_ok) {
+    auto container_client =
+        blob_service_client_->GetBlobContainerClient(location.container);
+    Azure::Storage::Blobs::ListBlobsOptions options;
+    options.Prefix = internal::EnsureTrailingSlash(location.path);
+    // https://learn.microsoft.com/en-us/rest/api/storageservices/blob-batch#remarks
+    //
+    // Only supports up to 256 subrequests in a single batch. The
+    // size of the body for a batch request can't exceed 4 MB.
+    const int32_t kNumMaxRequestsInBatch = 256;
+    options.PageSizeHint = kNumMaxRequestsInBatch;
+    try {
+      auto list_response = container_client.ListBlobs(options);
+      if (!missing_dir_ok && list_response.Blobs.empty()) {
+        return Status::IOError("Specified directory doesn't exist: ", location.path, ": ",
+                               container_client.GetUrl());
+      }
+      while (list_response.HasPage() && !list_response.Blobs.empty()) {
```

Comment:
This could be me being overly cautious, but maybe we shouldn't stop when `Blobs.empty()`, but instead run `continue;` and evaluate `MoveToNextPage()` again?

Reply:
Another thing: instead of `while`, you can use `for (; list_response.HasPage(); list_response.MoveToNextPage()) {` here.

Reply:
Both of `HasPage()` and `!Blobs.empty()` checks are needed to avoid empty `SubmitBatch()` request.
If there are no blobs under the given path, `ListBlobs()` returns `HasPage() == true` and `Blobs.empty() == true` response.

I'll use `for` as you suggested. I also found an azure-sdk-for-cpp document that uses `for`: https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/storage/MigrationGuide.md#listing-blobs-in-a-container

Reply:
> ListBlobs() returns HasPage() == true and Blobs.empty() == true response.

Is there a chance of that not being the last page? I see that you now only `continue;` instead of breaking the entire loop which I think is the more robust thing to do. 👍

Reply:
(Hmm. I think that the case (empty blob but it's not the last response) isn't happen but I'm not sure because Azure Blob Storage is a proprietary product...)